slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,316 @@
+//  PodE2EStartupLatency metric
+func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
+	controller := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				obj, err := pm.kubeClient.Core().Pods(NamespaceSystem).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return pm.kubeClient.Core().Pods(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Pod{},
+		func(obj interface{}) error {
+			p, ok := obj.(*v1.Pod)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Pod: %v", obj)
+				return nil
+			}
+			pm.handlePodUpdate(p)
+			return nil
+		},
+		func(obj interface{}) error {
+			p, ok := obj.(*v1.Pod)
+			if !ok {
+				if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+					p, ok = d.Obj.(*v1.Pod)
+					if !ok {
+						glog.Errorf("Failed to cast embedded object from tombstone to *v1.Pod: %v", d.Obj)
+						return nil
+					}
+				} else {
+					glog.Errorf("Failed to cast observed object to *v1.Pod: %v", obj)
+					return nil
+				}
+			}
+			pm.handlePodDelete(p)
+			return nil
+		},
+	)
+	go controller.Run(stopCh)
+
+	eventSelector := fields.Set{
+		"involvedObject.kind": "Pod",
+		// this is actually source.Component
+		"source": "kubelet",
+	}.AsSelector().String()
+
+	eventController := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = eventSelector
+				obj, err := pm.kubeClient.Core().Events(NamespaceSystem).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = eventSelector
+				return pm.kubeClient.Core().Events(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Event{},
+		func(obj interface{}) error {
+			e, ok := obj.(*v1.Event)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Event: %v", obj)
+				return nil
+			}
+			pm.handleEventUpdate(e)
+			return nil
+		},
+		func(obj interface{}) error {
+			return nil
+		},
+	)
+	go eventController.Run(stopCh)
+	wait.Until(pm.purgeOutdated, 15*time.Second, stopCh)
+	return nil
+}
+
+func (pm *PodStartupLatencyDataMonitor) purgeOutdated() {
+	now := time.Now()
+	pm.Lock()
+	defer pm.Unlock()
+	if len(pm.toDelete) == 0 {
+		return
+	}
+	left := []string{}
+	for _, podKey := range pm.toDelete {
+		if data, ok := pm.PodStartupData[podKey]; ok {
+			if data.deletedTimestamp.Add(pm.purgeAfter).After(now) {
+				left = append(left, podKey)
+				continue
+			}
+			delete(pm.PodStartupData, podKey)
+		}
+	}
+
+	pm.toDelete = left
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodUpdate(p *v1.Pod) {
+	if isReady, create, running := checkPodAndGetStartupLatency(p); isReady {
+		go pm.insertPodRunningTime(getPodKey(p), create, running, len(p.Spec.Containers))
```

Comment:
What will kubelet do if we have say 3 containers with exactly the same images? (like we do in kubemark)? Will there be 3 pulling/pulled events? Or only one?
If the latter, the len(containers) is probably wrong.